insights/v1alpha1/types_insights.go

```diff
@@ -0,0 +1,247 @@
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "ClearText" and "ObfuscateNetworking".
+	// When set to ClearText the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is ClearText.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy"`
+	// gatherersConfig is a list of gatherers configurations.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// +optional
+	GatherersConfig []GathererConfig `json:"gatherersConfig"`
+}
+
+const (
+	// No data obfuscation
+	NoPolicy DataPolicy = "ClearText"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicy = "ObfuscateNetworking"
+	// Data gathering is running
+	Running DataGatherState = "Running"
+	// Data gathering is completed
+	Completed DataGatherState = "Completed"
+	// Data gathering failed
+	Failed DataGatherState = "Failed"
+	// Data gathering is pending
+	Pending DataGatherState = "Pending"
+	// Gatherer state marked as disabled, which means that the gatherer will not run.
+	Disabled GathererState = "Disabled"
+	// Gatherer state marked as enabled, which means that the gatherer will run.
+	Enabled GathererState = "Enabled"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";ClearText;ObfuscateNetworking
+type DataPolicy string
+
+// state declares valid gatherer state types.
+// +kubebuilder:validation:Enum="";Enabled;Disabled
+type GathererState string
+
+// gathererConfig allows to configure specific gatherers
+type GathererConfig struct {
+	// name is the name of specific gatherer
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// state allows you to configure specific gatherer. Valid values are "Enabled", "Disabled" and omitted.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default.
+	// The current default is Enabled.
+	// +optional
+	State GathererState `json:"state"`
+}
+
+// dataGatherState declares valid gathering state types
+// +kubebuilder:validation:Optional
+// +kubebuilder:validation:Enum=Running;Completed;Failed;Pending
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Running' && self == 'Pending') && !(oldSelf == 'Completed' && self == 'Pending') && !(oldSelf == 'Failed' && self == 'Pending') && !(oldSelf == 'Completed' && self == 'Running') && !(oldSelf == 'Failed' && self == 'Running')", message="state cannot be changed backwards"
```

Comment:
Have you considered separating this into separate rules? Each individual rule is required to pass so you could split this
```suggestion
// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Running' && self == 'Pending')", message="state cannot transition from Running to Pending"
// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Completed' && self == 'Pending')", message="state cannot transition from Completed to Pending"
// +kubebuilder:validation:XValidation:rule='!(oldSelf == 'Failed' && self == 'Pending')", message="state cannot transition from Failed to Pending"
// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Completed' && self == 'Running')", message="state cannot transition from Completed to Running"
// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Failed' && self == 'Running')", message="state cannot transition from Failed to Running"
```

Reply:
Updated.